[image-picker][Android] Migrate to `registerForActivityResult`

bbarthec opened the pull request:
# Why

Migrates `image-picker` to take advantage of `registerForActivityResult`:
 - #17572

Partially resolves [ENG-3071](https://linear.app/expo/issue/ENG-3071/audit-expo-image-picker-android-rewrite-to-sweet-api)
Extracted part of #16251

# How

Main milestones from this PR:
- migration to use `Contracts` and `registerForActivityResult` with two callbacks
- simplified, coroutine-based logic behind each of the exported method
  - accessing optional metadata (exif, base64) done lazily
- limiting the state of the module to the single structure `pendingMediaPickingResult`

Things that changed tremendously:
- `ImagePickerOptions` is now `Serializable` to allow passing it as a additional parameter to `fallbackCallback` in `registerForActivityResult` mechanism
- `MediaHandler` alongside with `RawImageExporter` and `CompressionsImageExporter` are written in coroutines-based way with simple top-level flow
  - `ImageExporter` is simplified way to persist image data on disk
  - `ImageExportResult` is a class holding info about the exported (written to disk) image with lazy methods allowing accessing `exif` or `base64` data
- `ImageLibraryContract`, `CameraContract` and (_currently unused_) `CropImageContract` are contracts specifying how `registerForActivityResult` should be fired and how it should parse the results
- `ImagePickerModule` is simplified as much as possible - it only handles top-level logic of each exported method 

# Test Plan

- Open `native-component-list#ImagePicker` screens and play with the available methods

# Things to be delivered in separate PRs

- [ ] `editing` (aka `cropping`) images is now disabled as I need to provide some clever mechanism that would outlive possible Activity destruction
  - I can go with `registerForActivityResult`, but assuming that Android would kill the Application's main `Activity` every time some 3rd party `Activity` is called that would cause scenario like this
    - `1.` launch image picking (camera or image library) -- `Activity is killed` -->
    - `2.` pick an image -- `Activity is recreated and awaits for results (the whole Application is recreated, RN starts to render the App)` -->
    - `3.` `expo-image-picker` launches `CroppingActivity` -- `Activity is killed` -->
    - `4.` cropping operation -- `App is recreated` -->
    - `5.` results are delivered
    - ⚠️ we can see that the App is being recreated twice and that takes a lot of time
  - alternatively I can create custom `Activity` that would allow such scenario:
    - `1.` launch image picker with editing option enabled -- `Activity is killed` -->
    - `2.` custom Activity handles both picking and editing -- `Activity is recreated` -->
    - `3.` results are delivered
    - ✅ the main `Activity` is recreated only once
    - I would go with https://github.com/CanHub/Android-Image-Cropper that exposes `Activity` that allows both picking and cropping
- [ ] `multiselection` - there're some things in this PR that took into consideration multiselection, but it has to be delivered once `iOS` part will be delivered as `iOS` would lead the way and possible behaviour in edge-cases

bbarthec committed:
[image-picker][Android] Unify all exceptions

bbarthec committed:
[image-picker][Android] `./gradlew spotlessApply`

bbarthec committed:
Scope `ImagePickerFileProvider`

bbarthec committed:
Create `CameraContract`, `ImageLibraryContract` and `CropImageContract`

Contracts are responsible for handling asynchronous operation and delivering
the unified result.

bbarthec committed:
Add helper methods to `ImagePickerOptions`

Make `ImagePickerOptions` be responsible for transforming
itself to the necessary data type.

bbarthec committed:
Create top-level `MediaHandler`

`MediaHandler` is responsible for transforming results returned from contracts
into valid response ready to be delivered to JS.

bbarthec committed:
Adjust `ImageExporter` and it's implementations

- Get rid of `CropImageExporter` as it's no longer needed
- Make `ImageExporter` return `ImageExportResult` that can be used to read
additional info about the asset, like Exifs or Base64 data
- Refactor exporting process to be suspendable (coroutines)

bbarthec committed:
Get rid of `FileProvider`s

It's being replaces by simple `generateOutputPath` util function

bbarthec committed:
Remove old `Image/VideoResultTask`

It's being replaced by coroutines and simpler flow orchestrated by `MediaHandler`

bbarthec committed:
Remove `PickerResultsStore` and `PendingPromise`

These are replaced by `registerForActivity` mechanism that also allows
tracking launching Activity destruction

bbarthec committed:
Adjust `Constants` and `Exceptions`

bbarthec committed:
Refactor `ImagePickerUtils`

Uses heavily extensions functions

bbarthec committed:
Refactor `ImagePickerModule`

I've delegated as much as possible to other classes.
This class only registers the exported methods and handles very high level flow of each method.

It's using following mechanisms:
- `Coroutine`-backed exported methods
- `MediaHandler` for convering bare uri-based picked assets into JS ready response
- `launchForActivityResult` to propoerly call 3rd party Activities and monitors they possible destruction
  - the only state that is actually stored in this class narrows down to `PendingMediaPickingResult` that might come in handy whenever the launching/hosting Activity is destroyed

bbarthec committed:
[image-picker][Android] Adjustments after rebasing

bbarthec committed:
[image-picker][Android] Fix cancellation path

bbarthec committed:
[image-picker][Android] Restore `videoMaxDuration` for shooting video

bbarthec committed:
[image-picker][Android] Add support for videos

bbarthec committed:
Rename functions `toFileExtension ➡️ toImageFileExtension`

bbarthec committed:
Preserve Exif data across compress and edit operations

bbarthec committed:
Gracefuly handle video metadata extraction error

bbarthec committed:
[image-picker] Make `ImagePickerOptions` implement `Serializable` and migrate to `ContractOptions`

bbarthec committed:
[image-picker] Refactor `*Contract`s to be more stateless

bbarthec committed:
[image-picker] Refactor `ImagePickerModule`

- support new `registerForActivityResult` two-callbacks styled API
- temporarily disable cropping

bbarthec committed:
[image-picker] `./gradlew :expo-image-picker:spotlessApply`

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerUtils.kt:
```diff
@@ -1,91 +1,170 @@
-
-fun uriFromFilePath(path: String) = uriFromFile(File(path))
+internal fun String.toImageFileExtension(): String = when {
+  this.contains("png") -> ".png"
+  this.contains("gif") -> ".gif"
+  this.contains("bmp") -> ".bmp"
+  !this.contains("jpeg") -> {
```
```suggestion
  this.endsWith("png", ignoreCase = true) -> ".png"
  this.endsWith("gif", ignoreCase = true) -> ".gif"
  this.endsWith("bmp", ignoreCase = true) -> ".bmp"
  !this.endsWith("jpeg", ignoreCase = true) -> {
```
I think it should be done this way, cause someone can have a file called `png.gif` 😅

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerUtils.kt:
```diff
@@ -1,91 +1,170 @@
-    if (uri == newUri) {
-      return
+internal fun String.toBitmapCompressFormat(): Bitmap.CompressFormat = when {
+  this.contains("png") ||
+    this.contains("gif") ||
+    this.contains("bmp") -> {
```
We should use `endsWith(*, ignoreCase = true)`

bbarthec committed:
[image-picker] Replace `contains` with `endsWith(***, ignoreCase = true)`

Make checking file type more accurate by matching against filename sufix instead of the whole filename.

bbarthec committed:
[image-picker] Improve warning message
